core/range_scan_orchestrator.cxx

```diff
@@ -0,0 +1,420 @@
+      , num_vbuckets_{ num_vbuckets }
+      , scope_name_{ std::move(scope_name) }
+      , collection_name_{ std::move(collection_name) }
+      , scan_type_{ std::move(scan_type) }
+      , options_{ std::move(options) }
+      , vbucket_to_snapshot_requirements_{ mutation_state_to_snapshot_requirements(options_.consistent_with) }
+    {
+    }
+
+    auto scan()
+    {
+        range_scan_continue_options continue_options{
+            options_.batch_item_limit, options_.batch_byte_limit, options_.batch_time_limit, options_.retry_strategy, options_.ids_only,
+        };
+        continue_options.batch_time_limit = std::chrono::seconds{ 10 };
+        for (std::uint16_t vbucket = 0; vbucket < gsl::narrow_cast<std::uint16_t>(num_vbuckets_); ++vbucket) {
+            auto stream = std::make_shared<range_scan_stream>(io_,
+                                                              agent_,
+                                                              vbucket,
+                                                              range_scan_create_options{
+                                                                scope_name_,
+                                                                collection_name_,
+                                                                scan_type_,
+                                                                options_.timeout,
+                                                                {},
+                                                                vbucket_to_snapshot_requirements_[vbucket],
+                                                                options_.ids_only,
+                                                                options_.retry_strategy,
+                                                              },
+                                                              continue_options);
+            streams_[vbucket] = stream;
+            stream->start();
+        }
+
+        return scan_result(shared_from_this());
+    }
+
+    auto next() -> std::future<std::optional<range_scan_item>> override
+    {
+        auto barrier = std::make_shared<std::promise<std::optional<range_scan_item>>>();
+        if (options_.sort == scan_sort::none) {
+            next_item(streams_.begin(), [barrier](std::optional<range_scan_item> item) { barrier->set_value(std::move(item)); });
+        } else {
+            next_item_sorted({}, streams_.begin(), [barrier](std::optional<range_scan_item> item) { barrier->set_value(std::move(item)); });
+        }
+        return barrier->get_future();
+    }
+
+    void next(utils::movable_function<void(range_scan_item, std::error_code)> callback) override
+    {
+        auto handler = [callback = std::move(callback)](std::optional<range_scan_item> item) mutable {
+            if (item) {
+                callback(std::move(item.value()), {});
+            } else {
+                callback({}, errc::key_value::document_not_found);
```

Comment:
Is there a better way to signal that the streaming is complete than returning a document_not_found error?

Reply:
I will create another code here, as it still internal object. Just for easier handling other errors.